include/convex_bodies/hpolytope.h

```diff
@@ -638,6 +639,213 @@ class HPolytope {
+    {
+        if (lamda1 == lamda2)
+        {
+            return lamda1;
+        }
+        NT lamda;
+        const double tol = 1e-10;
+        if (lamda1 * lamda2 < NT(0)) 
+        {
+            if (previous_facet == current_facet) 
+            {
+                if (std::max(lamda1, lamda2) < NT(tol)) 
+                {
+                    lamda = std::min(lamda1, lamda2);
+                } 
+                else 
+                {
+                    lamda = std::max(lamda1, lamda2);
+                }
+            } else {
+                lamda = std::max(lamda1, lamda2);
+            }
+        } 
+        else 
+        {
+            if (previous_facet == current_facet) 
+            {
+                if (std::min(lamda1, lamda2) >= NT(0) && std::min(lamda1, lamda2) < NT(tol)) 
+                {
+                    lamda = std::max(lamda1, lamda2);
+                } 
+                else 
+                {
+                    lamda = std::min(lamda1, lamda2);
+                }
+            } 
+            else 
+            {
+                lamda = std::min(lamda1, lamda2);
+            }
+        }
+        return lamda;
+    }
+
+
+    //------------oracle for exact hmc spherical gaussian sampling---------------//
+
+    // compute intersection point of ray starting from r and pointing to v
+    // with polytope discribed by A and b
+    std::pair<NT, int> trigonometric_positive_intersect(Point const& r, Point const& v,
+                                                        NT const& omega, int &facet_prev) const
+    {
+
+        NT lamda = 0, C, Phi, t1, t2, tmin;
+        NT min_plus = std::numeric_limits<NT>::max(), t = std::numeric_limits<NT>::max();
+        NT max_minus = std::numeric_limits<NT>::lowest();
+        VT sum_nom, sum_denom;
+        unsigned int j;
+        int m = num_of_hyperplanes(), facet = -1;
+
+
+        sum_nom.noalias() = A * r.getCoefficients();
+        sum_denom.noalias() = A * v.getCoefficients();
+
+        NT* sum_nom_data = sum_nom.data();
+        NT* sum_denom_data = sum_denom.data();
+        const NT* b_data = b.data();
+
+        for (int i = 0; i < m; i++) {
+
+            C = std::sqrt((*sum_nom_data) * (*sum_nom_data) + ((*sum_denom_data) * (*sum_denom_data)) / (omega * omega));
+            Phi = std::atan((-(*sum_denom_data)) / ((*sum_nom_data) * omega));
+            if ((*sum_denom_data) < 0.0 && Phi < 0.0) {
+                Phi += M_PI;
+            } else if ((*sum_denom_data) > 0.0 && Phi > 0.0) {
+                Phi -= M_PI;
+            }
+
+            if (C > (*b_data)) {
+
+                t1 = (std::acos((*b_data) / C) - Phi) / omega;
+                if (facet_prev == i && std::abs(t1) < 1e-10){
+                    t1 = (2.0 * M_PI) / omega;
+                }
+                    
+                t2 = (-std::acos((*b_data) / C) - Phi) / omega;
```

Comment:
`std::acos((*b_data) / C)` can be precomputed

Reply:
done